### DISCUSSION TITLE
V5: setup and input functions

### DISCUSSION THREAD

**Discussion by christophe-g** [statelyai/xstate#4487] [Q&A] [answered]
Hey 

I am very new to xstate and loving it so far. 

One question re the new `setup` paradigm and ways to proceed with input functions. 

This is all very good: 
```js
const getChatCompletion = fromPromise(async () => { ... });

const machine = setup({
  actors: {
    getChatCompletion
     }
}).createMachine({
  states: {
    thinking: {
      invoke: { src: 'getChatCompletion', onDone: 'done' }
    },
    done: { type: 'final' }
  }
});
```

What about this - more precisely, is there a way to define the `input` function at setup stage:  
```js
const getChatCompletion = fromPromise(async (input) => { ... });

const machine = setup({
  actors: {
    getChatCompletion
    }
}).createMachine({
  states: {
    thinking: {
      invoke: { 
         src: 'getChatCompletion', 
         input: (context) => context.chatID,
         onDone: 'done' }
    },
    done: { type: 'final' }
  }
});
```

**Comment by Andarist** [comment, marked as answer]
> What about this - more precisely, is there a way to define the input function at setup stage:

No. `input` is something that should be passed to the actor per invocation. It's like a function argument - whereas within `setup` you configure available "functions".

**Reply by christophe-g** [reply]
Right - thanks a lot for the rapid answer !

This means machine definitions requiring `input` are not serializable, correct ?

**Reply by Andarist** [reply]
> This means machine definitions requiring input are not serializable, correct ?

Yes. They might not be fully serializable - but that was always kinda the case with a couple of features that were already available pre-input (like context factories). 

Does it create a problem for you? I'd love to get a better understanding of a use case for which this does matter.

**Reply by christophe-g** [reply]
> I'd love to get a better understanding of a use case for which this does matter.

Answering this with pleasure.

We are currently migrating our home-made FSM engine to xstate. The current implementation allows 2 different types of machines: 1. app level machines and 2. user-defined machines - stored in a db - for allowing clients to devise their own workflow (or other types of user-defined processes). 

Serialization is important for 2. 

The current machine engine is hosted on firebase, with cloud functions being triggered on transitions and handling actions. 

It is working quite well, but reason for migrating: 
- our implementation is not type safe 
- we are anyway migrating between 2 different firebase db, forcing some kind of code refactoring - this is an opportunity to revisit our FSM approach
- our design tool is nowhere as clean as stately.

**Reply by Andarist** [reply]
How do u deal with the serialization of what `setup` currently can "provide"? What does it mean that a user can define a machine and how do you deal with this later on? What kind of things do you supply on their behalf when you have to execute their workflow?

**Reply by christophe-g** [reply]
> How do u deal with the serialization of what setup currently can "provide"? 

For the time being - only migrating app level machine; the serialization will be a problem when we migrate user define machines. 

> What does it mean that a user can define a machine and how do you deal with this later on? 

A user has an interface (stately-, but still with a layout engine based on graphviz ) that allows him/her to define machine definition: 

For example - here a one stage review process:
```json
{
  "initialState": "started",
  "states": {
    "reviewed": {
      "finalType": "success",
      "stateType": "final",
      "title": "Reviewed"
    },
    "reviewing": {
      "actions": {
        "-LjyVsZb15dgDgDMLbpw": {
          "assignRule": "byRole",
          "createMachineID": "todo",
          "role": "steward",
          "transitionID": "reviewingReviewed",
          "type": "createMachine"
        }
      },
      "title": "Reviewing"
    },
    "started": {
      "actions": {
        "-LjyVsZb15dgDgDMLbpx": {
          "transitionID": "startedReviewing",
          "type": "makeTransition"
        }
      },
      "title": "Started"
    }
  },
  "title": "one stage review",
  "transitions": {
    "reviewingReviewed": {
      "from": "reviewing",
      "to": "reviewed"
    },
    "startedReviewing": {
      "from": "started",
      "to": "reviewing"
    }
  },
  "type": "review"
}
```

Machine instance exist live when there is an instance of this definition also stored on the db (with its context).  We allow a finite set of actions (e.g. create/instantiate another machine, create an new object, send a message, ...). It is possible to trigger a transition (similar to xstate `actor.send()`) by writing to a specific path in the db, similarly, we can create/launch a machine by writing to another path). 

> What kind of things do you supply on their behalf when you have to execute their workflow?

As we only support certain types of predefined actions, we support the implementation of those actions. They are triggered either on transition or state `entry`, depending on machine definition. Actions implementation have access to the current context and active state. 

Our view while designing this system was that machine definition was just data (e.g. serializable) that the engine should interpret. The way I read (but again, very new to the library) invoked actors `input` functions (in contrast with input values) is more as a contract between the `context` and the `actor` (and hence could potentially belong to `setup`), rather than 
> something that should be passed to the actor per invocation

I will also look into context factories, but imho having a serializable machine definition is quite a deal in general, not just for my small use case. 

And anyway, thanks for your incredible work here !

**Reply by Andarist** [reply]
> The way I read (but again, very new to the library) invoked actors input functions (in contrast with input values) is more as a contract between the context and the actor (and hence could potentially belong to setup), rather than

`input` could also depend on other things - like `event` and more. So it's not about about `context`. And some of those things could also be invocation-dependent. While maybe you don't use it that often you can actually `invoke`/`spawn` many instances of the same machine within a single parent. That's why we didn't bake this into `setup`. In a similar fashion actions and guards can be parameterized and the same action can be used multiple times through a machine but with different parameters:
```ts
entry: { type: 'reset', params: () => 100 },
exit: { type: 'reset', params: () => 0 }
```

So we followed the same principle with `input` here. It's also worth noting that in the past you could dynamically~ return a machine based on context/event/etc and in those situations, those acted kinda as an informal `input`. We had to make this impossible though since to pull off rehydration we need to know the `src` of an actor upfront - without evaluating any extra functions that could depend on values that we don't have access to at the time of rehydration.

We also had referenceable `input`s just up until recently (see this PR that removed it: https://github.com/statelyai/xstate/pull/4451 ) but it feels quite awkward. It also made it close to impossible to implement the required `input`s properly (see https://github.com/statelyai/xstate/pull/4463 ) because there were just 2 ways of providing that input and determining if it was definitely provided using the "other way" was rather hard.

I know that serialization in general is a desired feature but we also have quite a few things in the API that can be functions and having "referencable" slots in `setup` for all of them creates a cluttered API. 

Luckily, the machine definition is just an object... even if the final product might have to use plain functions for some things (like when it comes to the discussed `input`) you can always create your own lightweight serialization/deserialization mechanism. The `input` function itself isn't serializable at all anyway - so you need to keep some form of it on ur client anyway, right? So you could always crawl the received definition and replace a custom string reference with that function.

Don't get me wrong - any step further into supporting full serialization is something that we are willing to take. It's just that we need to account for some other desired requirements of the whole thing - like the API shape, its TS support, etc. If you have any ideas on how to cover both serialization and the other requirements then we'd gladly brainstorm them together with you.

**Reply by christophe-g** [reply]
> input could also depend on other things - like event and more

I was just about to comment on the fact that I realized this while digging into the examples ... to avoid you having to explain ! 

Thanks for this great answer.
